### PR TITLE
Fix --build_event_publish_all_actions with --remote_download_minimal

### LINKED ISSUE #10971: --remote_download_minimal does not work with --build_event_publish_all_actions
### Description of the problem / feature request:

--remote_download_minimal does not work with --build_event_publish_all_actions.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.
```
$ cat BUILD 
cc_binary(
    name = "test",
    srcs = ["test.cc"],
)
$ cat test.cc 
int main() {
    return 1;
}
```
```
# First, make sure that all actions are populated in the remote cache
# I'm using buildbarn, but I don't think this is related to the remote cache implementation.
bazel build ... --remote_cache=grpc://127.0.0.1:8980 --remote_instance_name=local --remote_upload_local_results=true
# Then, clean to ensure that remote_download_minimal is kicking in.
bazel clean
# Observe a failure from writing build events.
bazel build ... --remote_cache=grpc://127.0.0.1:8980 --remote_instance_name=local --remote_upload_local_results=true --remote_download_minimal --build_event_json_file=$PWD/be.json --build_event_publish_all_actions
INFO: Invocation ID: 87151a7b-6ca0-4c26-9ced-b4534bd969e9
INFO: Analyzed target //:test (13 packages loaded, 47 targets configured).
INFO: Found 1 target...
INFO: Deleting stale sandbox base /mnt/data2/cache/bazel/_bazel_lpeltonen/1da3da6443024d4847298a76d6f71841/sandbox
Target //:test up-to-date:
  bazel-bin/test
INFO: Elapsed time: 3.285s, Critical Path: 0.98s
INFO: 2 processes: 2 remote cache hit.
ERROR: Unable to write all BEP events to file due to 'java.io.FileNotFoundException: /mnt/data2/cache/bazel/_bazel_lpeltonen/1da3da6443024d4847298a76d6f71841/execroot/__main__/bazel-out/k8-fastbuild/bin/_objs/test/test.pic.o (No such file or directory)'
INFO: Build completed successfully, 5 total actions
```

### What operating system are you running Bazel on?

Ubuntu 18.04.

### What's the output of `bazel info release`?

release 2.2.0

Also reproduces on recent master (3c89a910ad1c59160ba4058c47b6752ce9e7efb6).

### What's the output of `git remote get-url origin ; git rev-parse master ; git rev-parse HEAD` ?

```
git@github.com:bazelbuild/bazel.git
3c89a910ad1c59160ba4058c47b6752ce9e7efb6
3c89a910ad1c59160ba4058c47b6752ce9e7efb6
```

###  Have you found anything relevant by searching the web?

No.

### Any other information, logs, or outputs that you want to share?

I think I understand what the issue is, I have opened #10972 that shows one way to fix the issue.

### PR DESCRIPTION
**Pull request by scele**
The ActionExecutedEvent references the primaryOutput of the executed
action (see ActionExecutedEvent::referencedLocalFiles).
ByteStreamBuildEventArtifactUploader::upload will try to ensure that
this referenced file is present on the remote cache.  It calls
file.isDirectory(), which will eventually go to
RemoteActionFileSystem::getRemoteInputMetadata.  However,
RemoteActionFileSystem::inputArtifactData only contains metadata about
the input files, not action outputs.

This change proposes a fix, where we make RemoteActionFileSystem
implement MetadataInjector, and inject information about output files
from ActionMetadataHandler::injectRemoteFile.

The implementation is likely not correct for all cases, but demonstrates
the issue and a possible approach to fix it.

Fixes #10971

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java [hunk 1]
```diff
@@ -339,6 +347,10 @@ private RemoteFileArtifactValue getRemoteInputMetadata(String execPathString) {
     if (m != null && m.isRemote()) {
       return (RemoteFileArtifactValue) m;
     }
+    RemoteFileArtifactValue rm = injectedMetadata.get(execPathString);
+    if (rm != null) {
+      return rm;
+    }
```
[the remark this answers is not in the document]
Comment (scele) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java]: Sure! Done.

**Comment by scele** [comment]
I need for apologies @janakdr, I know we're all busy. :) Thanks for the update, and the initial review!

**Comment by janakdr** [comment]
As I mentioned above, the Google-internal version of this filesystem injects metadata earlier, while the Spawn is running. The equivalent in open source would be somewhere in RemoteCache.java, I think, inside the RemoteCache#download* methods. After a discussion, we think that is the right place for it here as well. The idea is that the filesystem should be the source of truth for the build: metadata should hit it before anywhere else.

Moreover, internally at least, there are cases where the same Bazel "action" runs two commands. In that case, the second command consumes files produced by the first one, and so those files must be available on the filesystem, but those intermediate files are not considered action outputs, and so aren't ever injected into the MetadataHandler.

Finally, it would be great to have a test for the situation you're fixing. Thanks!

**Comment by meisterT** [comment]
What's the state of this PR?

**Comment by scele** [comment]
Unfortunately I have not had time to rebase and add tests for this. I'll close this for now, and re-open when I get some time to spend on this again.

**Pull request closed without merging**